add tutorial for chatbot with rag (#2141)
opensearch-project/ml-commons: docs/tutorials/agent_framework/Chatbot_with_RAG.md

# Topic

> Agent Framework is an experimental feature released in OpenSearch 2.12 and not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/ml-commons/issues/1161).

> This tutorial doesn't explain what retrieval-augmented generation(RAG) is.

> LLM has some known limitation, one is the frozen world to the time when the LLM trained. 

Reviewer: ```suggestion
One of the known limitations of large language models (LLMs) is that their knowledge base only contains information up to the time when the LLMs were trained. 
```

> They have no knowledge about recent events and your internal data. 

Reviewer 2: They have no knowledge about recent events and your internal data.  --> They have no knowledge of recent events or your internal data.

> They have no knowledge about recent events and your internal data. 

Reviewer: ```suggestion
LLMs have no knowledge of recent events or your internal data. 
```

> We can solve this problem by using retrieval-augmented generation(RAG).

Reviewer 2: Retrieval-Augmented Generation (RAG)

> We can solve this problem by using retrieval-augmented generation(RAG).

Reviewer: ```suggestion
You can augment the LLM knowledge base by using retrieval-augmented generation (RAG).
```

> This tutorial explains how to build your own Chatbot with Agent Framework and RAG (use your OpenSearch index as knowledge base).

Reviewer 2: This tutorial explains how to build your own chatbot with the Agent Framework and RAG (using your OpenSearch index as a knowledge base).

> This tutorial explains how to build your own Chatbot with Agent Framework and RAG (use your OpenSearch index as knowledge base).

Reviewer: ```suggestion
This tutorial explains how to build your own chatbot using the Agent Framework and RAG, supplementing the LLM knowledge with information contained in OpenSearch indexes.
```

> Note: You should replace the placeholders with prefix `your_` with your own value

Reviewer 2: Note: You should replace the placeholders prefixed with `your_` with your own values.

> Note: You should replace the placeholders with prefix `your_` with your own value

Reviewer: ```suggestion
Note: Replace the placeholders that start with `your_` with your own values.
```

# Steps

## 0. Preparation

Follow step 0 and step 1 of [RAG_with_conversational_flow_agent](./RAG_with_conversational_flow_agent.md) to set up

> knowledge base index `test_population_data` which contains population data of US cities.

Reviewer 2: to set up the knowledge base

> knowledge base index `test_population_data` which contains population data of US cities.

Reviewer: ```suggestion
the `test_population_data` index to provide supplementary information to the LLM. The index contains population data of US cities.
```

> Note the embedding model id, then create another knowledge base `test_tech_news` which contains recent tech news.

Reviewer: ```suggestion
Note the embedding model ID, which you'll use in the next steps. 
```

> Create ingest pipeline:

Reviewer: ```suggestion
Create an ingest pipeline:
```

```
PUT /_ingest/pipeline/test_tech_news_pipeline
{
    "description": "text embedding pipeline for tech news",
    "processors": [
        {
            "text_embedding": {
                "model_id": "your_text_embedding_model_id",
                "field_map": {
                    "passage": "passage_embedding"
                }
            }
        }
    ]
}
```

> Create index `test_tech_news`:

Reviewer: ```suggestion
Next, create another index named `test_tech_news`, which contains recent tech news:
```

```
PUT test_tech_news
{
  "mappings": {
    "properties": {
      "passage": {
        "type": "text"
      },
      "passage_embedding": {
        "type": "knn_vector",
        "dimension": 384
      }
    }
  },
  "settings": {
    "index": {
      "knn.space_type": "cosinesimil",
      "default_pipeline": "test_tech_news_pipeline",
      "knn": "true"
    }
  }
}
```

Ingest data. 
- The first two documents are from Wikipedia([Apple Vision Pro](https://en.wikipedia.org/wiki/Apple_Vision_Pro), [LLaMA](https://en.wikipedia.org/wiki/LLaMA)).

> - The third one is from [Amazon Bedrock document](https://aws.amazon.com/bedrock/)

Reviewer: ```suggestion
- The third one is from [Amazon Bedrock documentation](https://aws.amazon.com/bedrock/)
```

{"index":{"_index":"test_tech_news"}}
{"passage":"Amazon Bedrock is a fully managed service that offers a choice of high-performing foundation models (FMs) from leading AI companies like AI21 Labs, Anthropic, Cohere, Meta, Stability AI, and Amazon via a single API, along with a broad set of capabilities you need to build generative AI applications with security, privacy, and responsible AI. Using Amazon Bedrock, you can easily experiment with and evaluate top FMs for your use case, privately customize them with your data using techniques such as fine-tuning and Retrieval Augmented Generation (RAG), and build agents that execute tasks using your enterprise systems and data sources. Since Amazon Bedrock is serverless, you don't have to manage any infrastructure, and you can securely integrate and deploy generative AI capabilities into your applications using the AWS services you are already familiar with."}

```
## 1. Create LLM

> Follow step 2 "Prepare LLM" of [RAG_with_conversational_flow_agent](./RAG_with_conversational_flow_agent.md) to set up Bedrock Claude model.

Reviewer 2: to set up the Bedrock Claude model.

> Follow step 2 "Prepare LLM" of [RAG_with_conversational_flow_agent](./RAG_with_conversational_flow_agent.md) to set up Bedrock Claude model.

Reviewer: ```suggestion
Follow step 2 (Prepare LLM) of [RAG_with_conversational_flow_agent](./RAG_with_conversational_flow_agent.md) to set up the Bedrock Claude model.
```

> Note the model id, you will use it in following steps.

Reviewer 2: Note the model ID; you will use it in the following steps.

> Note the model id, you will use it in following steps.

Reviewer: ```suggestion
Note the model ID, you will use it in following steps.
```

## 2. Create Agent
Create agent with `conversational` type. 

> Both `conversational_flow` and `conversational` agents can support conversation history.

Reviewer 2: Both conversational_flow and conversational agents can support conversation history, but they differ in their operation:

Difference:

> - `conversational_flow` agent runs tools sequentially with predefined order.

Reviewer 2: agent runs tools sequentially in a predefined order.

> - `conversational` agent dynamically chooses which tool to run.

Reviewer 2: agent dynamically chooses which tool to run next.

> This agent includes two tools which provides recent population data and tech news.

Reviewer 2: This agent includes two tools that provide recent population data and tech news, respectively.

Explanation:

> - `"max_iteration": 5`: Agent will run at most 5 times of LLM.

Reviewer 2: The agent will run the LLM a maximum of 5 times.

> - `"response_filter": "$.completion"` is to filter only the answer from Bedrock Claude model response

Reviewer 2: is used to filter only the answer from the Bedrock Claude model's response.

> - `"doc_size": 3` in `population_data_knowledge_base`: means it will return top 3 documents.

Reviewer 2: means it will return the top 3 documents.

  "llm": {
    "model_id": "your_llm_model_id",
    "parameters": {
      "max_iteration": 5,
      "response_filter": "$.completion"
    }
  },
  "memory": {
    "type": "conversation_index"
  },
  "tools": [
    {
      "type": "VectorDBTool",
      "name": "population_data_knowledge_base",
      "description": "This tool provides population data of US cities.",
      "parameters": {
        "input": "${parameters.question}",
        "index": "test_population_data",
        "source_field": [
          "population_description"
        ],
        "model_id": "your_text_embedding_model_id",
        "embedding_field": "population_description_embedding",
        "doc_size": 3
      }
    },
    {
      "type": "VectorDBTool",
      "name": "tech_news_knowledge_base",
      "description": "This tool provides recent tech news.",
      "parameters": {
        "input": "${parameters.question}",
        "index": "test_tech_news",
        "source_field": [
          "passage"
        ],
        "model_id": "your_text_embedding_model_id",
        "embedding_field": "passage_embedding",
        "doc_size": 2
      }
    }
  ],
  "app_type": "chat_with_rag"
}
```

> Note the agent id, you will use it in next step.

Reviewer 2: Note the agent ID; you will use it in the next step.

## 3. Test Agent

> `conversational` agent supports `verbose`. You can set it as `true` to see detail steps.

Reviewer 2: The `conversational` agent supports a `verbose` option. You can set it to `true` to see detailed steps.

> Or you can use get trace data API.

Reviewer 2: Alternatively, you can use the get trace data API.

```
GET _plugins/_ml/memory/message/message_id/traces
```

### 3.1 Start new conversation

> - Example1: ask question related to tech news

Reviewer 2: Example 1: Ask a question related to tech news.

```
POST _plugins/_ml/agents/your_agent_id/_execute
{
  "parameters": {
    "question": "What's vision pro",
    "verbose": true
  }
}
```

> From response, we can see Agent runs `tech_news_knowledge_base` tool and get top 2 documents. Agent will pass these documents as context to LLM.

Reviewer 2: From the response, we can see the Agent runs the `tech_news_knowledge_base` tool and retrieves the top 2 documents. The Agent then passes these documents as context to the LLM, allowing it to accurately answer the question.

"""
        },
        {
          "name": "response",
          "result": "Vision Pro is a mixed-reality headset developed by Apple that was announced in 2023. It uses cameras and sensors to overlay digital objects and information on the real world. The device runs an operating system called visionOS that allows users to interact with windows and apps in a 3D environment using gestures, eye tracking, and voice commands."
        }
      ]
    }
  ]
}
```

> You can also check detail steps in trace data:

Reviewer 2: You can also check detailed steps in the trace data:

```
GET _plugins/_ml/memory/message/ebVSxI0B8vrNLhb9nxty/traces
```

> - Example2: ask question related to city population

Reviewer 2: Ask a question

```
POST _plugins/_ml/agents/your_agent_id/_execute
{
  "parameters": {
    "question": "What's the population of Seattle 2023",
    "verbose": true
  }
}
```

> From response, we can see Agent runs `population_data_knowledge_base` tool and get top 3 documents. Agent will pass these documents as context to LLM.

Reviewer 2: From the response, we can see the Agent runs the `population_data_knowledge_base` tool and retrieves the top 3 documents. The Agent then passes these documents as context to the LLM, which allows it to accurately answer the question.

{"_index":"test_population_data","_source":{"population_description":"Chart and table of population level and growth rate for the New York City metro area from 1950 to 2023. United Nations population projections are also included through the year 2035.\\nThe current metro area population of New York City in 2023 is 18,937,000, a 0.37% increase from 2022.\\nThe metro area population of New York City in 2022 was 18,867,000, a 0.23% increase from 2021.\\nThe metro area population of New York City in 2021 was 18,823,000, a 0.1% increase from 2020.\\nThe metro area population of New York City in 2020 was 18,804,000, a 0.01% decline from 2019."},"_id":"AxF5vo0BubpYKX5ER0fT","_score":0.56461215}
"""
        },
        {
          "name": "response",
          "result": "According to the population data tool, the population of Seattle in 2023 is approximately 3,519,000 people, a 0.86% increase from 2022."
        }
      ]
    }
  ]
}
```

### 3.2 Continue a conversation
```
POST _plugins/_ml/agents/your_agent_id/_execute
{
  "parameters": {
    "question": "What's the population of Austin 2023, compare with Seattle",
    "memory_id": "l7VUxI0B8vrNLhb9sRuQ",
    "verbose": true
  }
}
```

> From response, you can see `population_data_knowledge_base` doesn't return Seattle population, but agent learnt from history messages

Reviewer 2: From the response, you can observe that the `population_data_knowledge_base` does not return the population of Seattle. However, the agent learns from historical messages that the population of Seattle is "approximately 3,519,000 people, a 0.86% increase from 2022."
